collectd/collectd.go

```diff
@@ -0,0 +1,220 @@
+	DefaultPort = 25826
+)
+
+var (
+	// ErrBindAddressRequired is returned when starting the Server
+	// without a TCP or UDP listening address.
+	ErrBindAddressRequired = errors.New("bind address required")
+
+	// ErrDatabaseNotSpecified retuned when no database was specified in the config file
+	ErrDatabaseNotSpecified = errors.New("database was not specified in config")
+
+	// ErrCouldNotParseTypesDBFile returned when unable to parse the typesDBfile passed in
+	ErrCouldNotParseTypesDBFile = errors.New("could not parse typesDBFile")
+
+	// ErrServerClosed return when closing an already closed graphite server.
+	ErrServerClosed = errors.New("server already closed")
+
+	// ErrResolveUDPAddr returned when we are unable to resolve a udp address
+	ErrResolveUDPAddr = errors.New("Unable to resolve UDP address")
+
+	// ErrListenUDP returned when we are unable to resolve a udp address
+	ErrListenUDP = errors.New("Unable to listen on UDP")
+)
+
+// SeriesWriter defines the interface for the destination of the data.
+type SeriesWriter interface {
+	WriteSeries(database, retentionPolicy, name string, tags map[string]string, timestamp time.Time, values map[string]interface{}) error
+}
+
+type Server struct {
+	mu   sync.Mutex
+	wg   sync.WaitGroup
+	done chan struct{} // close notification
+	conn *net.UDPConn
+
+	writer      SeriesWriter
+	Database    string
+	typesdb     gollectd.Types
+	typesdbpath string
+}
+
+func NewServer(w SeriesWriter, typesDBPath string) *Server {
+	s := &Server{}
+
+	s.writer = w
+	s.typesdbpath = typesDBPath
+	s.typesdb = make(gollectd.Types)
+
+	return s
+}
+
+func (s *Server) ListenAndServe(iface string) error {
+	if iface == "" { // Make sure we have an address
+		return ErrBindAddressRequired
+	} else if s.Database == "" { // Make sure they have a database
+		return ErrDatabaseNotSpecified
+	}
+
+	var err error
+	s.typesdb, err = gollectd.TypesDBFile(s.typesdbpath)
+	if err != nil {
+		return ErrCouldNotParseTypesDBFile
+	}
+
+	addr, err := net.ResolveUDPAddr("udp", iface)
+	if err != nil {
+		return ErrResolveUDPAddr
+	}
+
+	conn, err := net.ListenUDP("udp", addr)
+	if err != nil {
+		return ErrListenUDP
+	}
+	s.conn = conn
+
+	// Create a new close notification channel.
+	done := make(chan struct{}, 0)
+	s.done = done
+
+	s.wg.Add(1)
+	go s.serve(conn)
+
+	return nil
+}
+
+func (s *Server) serve(conn *net.UDPConn) {
+	defer s.wg.Done()
+
+	// From https://collectd.org/wiki/index.php/Binary_protocol
+	//   1024 bytes (payload only, not including UDP / IP headers)
+	//   In versions 4.0 through 4.7, the receive buffer has a fixed size
+	//   of 1024 bytes. When longer packets are received, the trailing data
+	//   is simply ignored. Since version 4.8, the buffer size can be
+	//   configured. Version 5.0 will increase the default buffer size to
+	//   1452 bytes (the maximum payload size when using UDP/IPv6 over
+	//   Ethernet).
+	buffer := make([]byte, 1452)
+
+	for {
+		select {
+		case <-s.done:
+			s.conn.Close()
```

Comment:
Closing `s.conn` should be done in `Server.Close()` otherwise your server will most likely hang while waiting for `conn.ReadFromUDP()` and never receive the signal from `s.done`.